pkg/apis/kubeflow.org/v2alpha1/trainjob_types.go

```diff
@@ -15,3 +15,235 @@ limitations under the License.
+	ModelConfig *ModelConfig `json:"modelConfig,omitempty"`
+
+	// Labels to apply for the derivative JobSet and Jobs.
+	// They will be merged with the TrainingRuntime values.
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// Annotations to apply for the derivative JobSet and Jobs.
+	// They will be merged with the TrainingRuntime values.
+	Annotations map[string]string `json:"annotations,omitempty"`
+
+	// Custom overrides for the training runtime.
+	PodSpecOverrides []PodSpecOverrides `json:"podSpecOverrides,omitempty"`
+
+	// Whether the controller should suspend the running TrainJob.
+	// Defaults to false.
+	Suspend *bool `json:"suspend,omitempty"`
+
+	// ManagedBy field indicates the controller that manages a TrainJob.
+	ManagedBy *string `json:"managedBy,omitempty"`
+}
+
+// TrainingRuntimeRef represents the reference to the existing training runtime.
+type TrainingRuntimeRef struct {
+	// Name of the runtime being referenced.
+	// When namespaced-scoped TrainingRuntime is used, the TrainJob must have
+	// the same namespace as the deployed runtime.
+	Name string `json:"name"`
+
+	// APIGroup of the runtime being referenced.
+	// Defaults to `kubeflow.org`.
+	APIGroup *string `json:"apiGroup,omitempty"`
+
+	// Kind of the runtime being referenced.
+	// It must be one of TrainingRuntime or ClusterTrainingRuntime.
+	// Defaults to ClusterTrainingRuntime.
+	Kind *string `json:"kind,omitempty"`
+}
+
+// Trainer represents the desired trainer configuration.
+// Every training runtime contains `trainer` container which represents Trainer.
+type Trainer struct {
+	// Docker image for the training container.
+	Image string `json:"image,omitempty"`
+
+	// Entrypoint commands for the training container.
+	Command []string `json:"command,omitempty"`
+
+	// Arguments to the entrypoint for the training container.
+	Args []string `json:"args,omitempty"`
+
+	// List of environment variables to set in the training container.
+	// These values will be merged with the TrainingRuntime's trainer environments.
+	Env []corev1.EnvVar `json:"env,omitempty"`
+
+	// Number of training nodes.
+	// TODO (andreyvelich): Do we want to support dynamic num of nodes in TrainJob for PyTorch elastic: `--nnodes=1:4` ?
+	NumNodes *int32 `json:"numNodes,omitempty"`
+
+	// Compute resources for each training node.
+	ResourcesPerNode corev1.ResourceRequirements `json:"resourcesPerNode,omitempty"`
+
+	// Number of processes/workers/slots on every training node.
+	// For the Torch runtime: `auto`, `cpu`, `gpu`, or int value can be set.
+	// For the MPI runtime only int value can be set.
+	NumProcPerNode *string `json:"numProcPerNode,omitempty"`
+}
+
+// DatasetConfig represents the desired dataset configuration.
+// When this API is used, the training runtime must have
+// the `dataset-initializer` container in the `Initializer` Job.
+type DatasetConfig struct {
+	// Storage uri for the dataset provider.
+	StorageUri string `json:"storageUri"`
```

Comment:
Could StorageUri be a pointer to the string? The validation could differentiate between unspecified and empty string. 

Also `omitempty` is still necessary? From a PATCH perspective, if this field is unspecified, its zero value will be included in the serialization.

Reply:
Let's make the storageUri optional given that Training Runtime can contain the default dataset and model.